apps/els_lsp/src/els_buffer_server.erl

```diff
@@ -0,0 +1,126 @@
+%%%=============================================================================
+%%% @doc Buffer edits to an open buffer to avoid re-indexing too often.
+%%% @end
+%%%=============================================================================
+-module(els_buffer_server).
+
+%%==============================================================================
+%% API
+%%==============================================================================
+-export([ new/2
+        , stop/1
+        , apply_edits/2
+        , flush/1
+        ]).
+
+-export([ start_link/2 ]).
+
+%%==============================================================================
+%% Callbacks for the gen_server behaviour
+%%==============================================================================
+-behaviour(gen_server).
```

Comment:
The buffer was introduced as part of [support for incremental text sync](https://github.com/erlang-ls/erlang_ls/pull/1002) and I just ported it to a `gen_server`, so it's not new with this feature.

The main intention was to buffer `didChange` events, so that re-indexing would be retriggered not too often (a didChange event is sent at every keystroke). But I agree that it complicates things and probably introduces some races.